c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1470,290 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+  // member class template of Coroutine<T>, it is able to implement an
+  // `await_suspend(Coroutine<T>::Handle)` override, providing it type-safe access to our enclosing
+  // coroutine's promise adapter. An `operator co_await()` free function would have to implement
+  // a type-erased `await_suspend(stdcoro::coroutine_handle<void>)` override, and implement
+  // suspension and resumption in terms of .then(). Yuck!
+
+private:
+  friend class CoroutineBase<Coroutine<T>, T>;
+
+  class Adapter;
+  Adapter* adapter = nullptr;
+  // Set by Adapter's ctor in `get_return_object()`. Nulled out again by Adapter's dtor, if the
+  // promise returned from `get_return_object()` is destroyed while this coroutine is suspended.
+};
+
+template <typename Self, typename T>
+class CoroutineBase {
+public:
+  void return_value(T value) {
+    static_cast<Self*>(this)->adapter->fulfiller.fulfill(kj::mv(value));
+  }
+};
+template <typename Self>
+class CoroutineBase<Self, void> {
+public:
+  void return_void() {
+    static_cast<Self*>(this)->adapter->fulfiller.fulfill();
+  }
+};
+// The Coroutines spec has no `_::FixVoid<T>` equivalent to unify valueful and valueless co_return
+// statements, and programs are ill-formed if the coroutine implementation object (Coroutine<T>) has
+// both a `return_value()` and `return_void()`. No amount of EnableIffery can get around it, so
+// these return_* functions live in a CRTP mixin.
+
+template <typename T>
+class Coroutine<T>::Adapter: private Event {
+public:
+  explicit Adapter(PromiseFulfiller<T>& f, Coroutine::Handle coroutine)
+      : fulfiller(f), coroutine(coroutine) {
+    coroutine.promise().adapter = this;
+  }
+  KJ_DISALLOW_COPY(Adapter);
+
+  ~Adapter() noexcept(false) {
+    // If a coroutine runs to completion, its frame will already have been destroyed by the time the
+    // adapter is destroyed, so this if-condition will almost always be false.
+    if (coroutine) {
+      // coroutine.promise() is a reference to our implementation object, Coroutine<T>.
+      coroutine.promise().adapter = nullptr;
+      coroutine.destroy();
+    }
+  }
+
+  void coroutineDestroyed() {
+    // The coroutine was destroyed before the adapter.
+    coroutine = nullptr;
+  }
+
+  void awaitSuspend(PromiseNode& nodeParam, ExceptionOrValue& resultParam) {
+    // Called by Awaiter::await_suspend() to tell us to wait for `node` to become ready, then store
+    // the node's result in `result`.
+
+    // We'll need to save these references for fire(). Note that Awaiter is movable, and `result` is
+    // its subobject, but it is safe to store a reference to `result` in Adapter, because by the
+    // time `await_suspend()` is called, Awaiter has already been returned from `await_transform()`
+    // and stored in its final resting place in the coroutine's frame.
+    node = &nodeParam;
+    result = &resultParam;
+
+    // We use the promise adapter as an Event, because we must be able to destroy the coroutine from
```

Comment:
Technically `fire()` callbacks _do_ have the capability to delete themselves, by returning an `Own<Event>` pointing to themselves, which the caller drops shortly thereafter. I suppose you'd need a custom disposer but it looks like it should work.

Would there be any advantage to doing it that way?

Reply:
True, a custom disposer didn't occur to me. I think it'd work, too. I think the main advantage would be that it would get rid of the pointer twiddling I'm doing here (obviating the source of my nullifyPointers() paranoia). The main disadvantage would be that the coroutine would run in a destructor, which would make some stack traces aesthetically irritating.

To be honest, I could go either way. I think the code would be a little simpler, the stack trace would be uglier, and I don't think there'd be any performance difference.

Reply:
Oh I see, `.resume()` can end up destroying the coroutine with no way to stop it, huh? That creates two problems:

1. We definitely can't be executing the coroutine in the destructor of the event returned by `fire()`. It's important that application code runs _during_ `fire()`.
2. This implies that my comment below about reducing the allocations from 2 to 1 by having `Coroutine` implement `PromiseNode` won't work.

Is this where it might make sense to change `final_suspend()` to `suspend_always()`? In that case I suppose we gain full control over exactly when the coroutine is destroyed? And then we'd have it happen strictly when the caller drops the `PromiseNode`, which seems like it simplifies a lot.

Reply:
> Is this where it might make sense to change final_suspend() to suspend_always()?

Yup, I think that will be required to support a PromiseNode implementation. I like the idea a lot -- the Promise will just be a pointer to an object on the coroutine frame. It also makes our terminology and the standard's terminology make more sense together ("a kj::Promise<T>'s promise_type is PromiseNode").

> We definitely can't be executing the coroutine in the destructor of the event returned by fire()

I think this means that the promise adapter (soon the PromiseNode) will need to remain the Event, rather than the Awaiters. While we can prevent `coroutine.resume()` from destroying the coroutine, we can't prevent it from destroying the Awaiter, which is essentially a temporary object kept alive for the duration of the suspension.

Reply:
Ah, yeah, I didn't ever imagine the awaiter would be the event... I imagined the `Coroutine<T>` would be the event. I think that's the plan now, right? `Coroutine<T>` will become the `PromiseNode` (subsuming the adapter's role) and therefore also the `Event`?

Reply:
Oh, I get it! Haha, yes, that's the plan now.